libs/s25main/EconomyModeHandler.cpp

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2005 - 2020 Settlers Freaks (sf-team at siedler25.org)
+//
+// This file is part of Return To The Roots.
+//
+// Return To The Roots is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Return To The Roots is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Return To The Roots. If not, see <http://www.gnu.org/licenses/>.
+
+#include "EconomyModeHandler.h"
+
+#include "EventManager.h"
+#include "GameInterface.h"
+#include "GamePlayer.h"
+#include "GlobalGameSettings.h"
+#include "SerializedGameData.h"
+#include "helpers/containerUtils.h"
+#include "helpers/make_array.h"
+#include "random/Random.h"
+#include "world/GameWorld.h"
+#include "world/GameWorldGame.h"
+#include "gameTypes/JobTypes.h"
+#include "gameData/JobConsts.h"
+#include <boost/optional.hpp>
+#include <array>
+
+EconomyModeHandler::EconomyModeHandler(unsigned endFrame) : endFrame(endFrame), gfLastUpdated(0)
+{
+    constexpr auto specialGoodPool =
+      helpers::make_array(GD_TONGS, GD_HAMMER, GD_AXE, GD_SAW, GD_PICKAXE, GD_SHOVEL, GD_CRUCIBLE, GD_RODANDLINE,
+                          GD_SCYTHE, GD_CLEAVER, GD_ROLLINGPIN, GD_BOW);
+
+    constexpr auto commonGoodPool =
+      helpers::make_array(GD_BEER, GD_WATER, GD_BOAT, GD_SWORD, GD_IRON, GD_FLOUR, GD_FISH, GD_BREAD, GD_WOOD,
+                          GD_BOARDS, GD_STONES, GD_GRAIN, GD_COINS, GD_GOLD, GD_IRONORE, GD_COAL, GD_MEAT, GD_HAM);
+
+    constexpr unsigned numGoodTypesToCollect = 7;
+
+    // Randomly determine *numGoodTypesToCollect* many good types, one of which is a special good (=tool)
+
+    static_assert(numGoodTypesToCollect > 0, "There have to be goods to be collected");
+    static_assert(commonGoodPool.size() >= numGoodTypesToCollect - 1, "There have to be enough commond goods");
+    static_assert(!specialGoodPool.empty(), "There have to be enough special goods");
+    goodsToCollect.clear();
+    goodsToCollect.resize(numGoodTypesToCollect);
```

Comment:
Just stumbled upon this and took a little bit to long to understand it although it should be very easy. May I suggest a minor change: Don't use `resize` but `reserve` (nit: The `clear` is superflous and can be removed), use a regular `for` loop with `push_back` below and instead of `std::find` you can use `helpers::contains`

That is if you want to keep using a vector here, the code is fully reasonable for an array. Was there a reason you switched this to vector although the `numGoodTypesToCollect` is a compile time constant?

Reply:
Using reserve and contains is a good change, thanks for the suggestion, did it. 

Going to a regular for loop would require changing the goods pools to sets, because I don't want duplicates in goodsToCollect. With the current implementation if the random generator gives out a good we already have in goodsToCollect I just go for another round in the while loop until I have found enough unique goods. If I changed the goods pools to sets I could just remove any found goods from the set so I wouldn't have to do extra loop iterations. The overhead didn't seem worth it to me though, especially since I asserted above that there are enough goods in the pool, so I won't run into an infinite loop. If you prefer the set approach for clarity, though, I will do it.

I changed goodsToCollect into a vector when I worked on it anyway, because then numGoodTypesToCollect can be nicely contained only in the constructor. This makes it possible to very easily have an addon option to change the number in the future, if you want. I didn't want to implement such an addon option for this pull request though, because it's not very important and I didn't want to put more features into this pull request (uncharacteristical for me, I know xD).

Reply:
Oh, now I remember, another reason to use a vector here, was that SerializedGameData::PopContainer at the moment only seems to support dynamically sized containers.